gms/gossiper.cc

```diff
@@ -1840,7 +1840,7 @@ future<> gossiper::start_gossiping(int generation_nbr, std::map<application_stat
     co_await container().invoke_on_all([] (gms::gossiper& g) {
         g._failure_detector_loop_done = g.failure_detector_loop();
     });
-    co_await _direct_fd_pinger.update_generation_number(generation_nbr);
+    co_await _echo_pinger.update_generation_number(generation_nbr);
```

Comment:
I think this could be made even simpler by specifying "echo_pinger" interface as a gossiper base class and passing a gossiper reference into direct_fd_pinger to do the pings. I don't quite get the point of having an echo_pinger in between. But  I don't regard this as a  critical comment.

Reply:
Personally I prefer this way. I don't like this composition-through-inheritance pattern.

Reply:
Agree with @kbr-. Sometimes using inheritance is important for performance, but usually it just obfuscates things.